// tracker/search/searchTask.js
var Q = require('q');
var _ = require('lodash');
var Moment = require('moment');
var Msg = require('../message');
var ObjectId = require('mongoose').Types.ObjectId;

//models
var Task = require('../model/Task');
var Log = require('../model/log');


exports.query = function (options) {
    options = _.extend({ populate: true }, options);
    if (options.id) {
        return queryTaskById(options);
    } else {
        return queryTasksByConditions(options);
    }
};


function queryTasksByConditions(options) {
    var deferred = Q.defer();
    var conditions = getQueryConditions(options);
    var queryOptions = getQueryOptions(options);
    var query = Task.find(conditions, null, queryOptions);
    console.log(JSON.stringify(conditions), JSON.stringify(queryOptions));
    if (options.populate) {
        query = query.populate([{
            path: 'children'
        }]);
    }
    var taskStatus = options.status;
    var isMarked = (options.marked === 'true' || options.marked === true);
    var withoutParent = (options.parent === undefined);

    query.exec(function (err, tasks) {
        if (err) {
            return deferred.reject(err);
        }
        console.log(tasks);
        if (_.isEmpty(tasks)) {
            return deferred.resolve([]);
        }
        tasks.forEach(function (task) {
            task.children = task.children.filter(function (childTask) {
                var bool = true;
                if (isMarked) {
                    bool = bool && childTask.marked === true;
                }
                if (taskStatus === 'doing') {
                    return  bool && childTask.progress > 0 && childTask.progress < 100;
                } else if (taskStatus === 'done') {
                    return  bool && childTask.progress === 100;
                } else {
                    return bool && true;
                }
            }).sort(function (a, b) {
                return new Date(b.lastActiveTime).getTime() - new Date(a.lastActiveTime).getTime();
            });
        });

        if (withoutParent) {
            tasks = tasks.filter(function (task) {
                var parent = task.parent;
                //task is not a subtask of the task in the `tasks`
                var result  = !parent || !tasks.some(function (task) {
                    return task._id.toString() === parent.toString();
                });
                return result;
            });
        }
        //calculate total time consume for each task
        if (options.calculateTimeConsume) {
            Q.allSettled(tasks.map(calculateTaskTimeConsume)).then(function () {
                deferred.resolve(tasks);
            });
        } else {
            deferred.resolve(tasks);
        }

        function calculateTaskTimeConsume(task) {
            var deferred = Q.defer();
            Log.aggregate()
                .match({task: new ObjectId(task._id)})
                .group({_id: '$task', totalTime: {$sum: "$len"}})
                .exec()
                .then(function (result) {
                    var calculteResult = result[0],
                        totalTime = 0;
                    if (!calculteResult) {
                        totalTime = 0;
                    } else {
                        totalTime = calculteResult.totalTime;
                    }
                    //if have children, then calculate the children's time consume
                    if (task.children.length > 0) {
                        Q.allSettled(task.children.map(calculateTaskTimeConsume)).then(function (results) {
                            totalTime = results.reduce(function (totalTime, result) {
                                if (result.state === "fulfilled") {
                                    totalTime += result.value;
                                }
                                return totalTime;
                            }, totalTime);
                            task.setValue('totalTime', totalTime);
                            deferred.resolve();
                        });
                    } else {
                        task.setValue('totalTime', totalTime);
                        deferred.resolve(totalTime);
                    }
                }).on('reject', function (err) {
                    deferred.reject(err);
                });
            return deferred.promise;
        }
    });

    return deferred.promise;
}

function queryTaskById(id) {

}


function getQueryConditions(options) {
    var parentCondition;
    if (options.parent === null || options.parent === 'null') {
        parentCondition =  {parent: { $type: 10 }};
    } else if (options.parent) {
        parentCondition = {parent: options.parent};
    }
    var conditions = [],
        timeCondition = [];
    if (parentCondition) {
        conditions.push(parentCondition);
    }
    if (options.start && options.end) {
        timeCondition.push({
            createTime: {
                $gt: new Moment(options.start).startOf('day').toDate(),
                $lt: new Moment(options.end).endOf('day').toDate()
            }
        });

        timeCondition.push({
            lastActiveTime: {
                $gt: new Moment(options.start).startOf('day').toDate(),
                $lt: new Moment(options.end).endOf('day').toDate()
            }
        });

        timeCondition.push({
            $and: [{
                lastActiveTime: { $gte: new Moment(options.end).endOf('day').toDate() }
            }, {
                createTime: {$lte: new Moment(options.start).startOf('day').toDate() }
            }]
        });
    }
    var taskStatus = options.status;
    if (taskStatus === 'doing') {
        conditions.push({
            $or: [
                {progress: {$gte: 0, $lt: 100}}, //doing
                {progress: -1, children : {$exists:true}, $where:'this.children.length>0'} //have no progress but have children
            ]
        });
    } else if (taskStatus === 'done'){
        conditions.push({
            $or: [{
                progress: 100
            }, {
                progress: {$lt: 100},
                children : {$exists:true},
                $where:'this.children.length>0'
            }]
        });
    }
    if (options.projectId) {
        conditions.push({
            projectId: new ObjectId(options.projectId)
        });
    }

    if (options.versionId) {
        conditions.push({
            versionId: new ObjectId(options.versionId)
        });
    }

    var isMarked = (options.marked === 'true' || options.marked === true);

    if (isMarked) {
        conditions.push({
            marked: isMarked
        });
    }

    if (options.name) {
        conditions.push({
            name: options.name
        });
    }

    if (!_.isEmpty(timeCondition)) {
        conditions.push({
            $or: timeCondition
        });
    }


    if (options.dueDays) {
        var dueDays = parseInt(options.dueDays, 10);
        //var now = new Moment();
        //console.log(dueDays, new Moment().add(dueDays, 'days').format('YYYY-MM-DD'));
        conditions.push({
            dueTime: {
                "$gt": new Moment().toDate(),
                "$lte": new Moment().add(dueDays, 'days').toDate()
            }
        });
    }

    if (!_.isEmpty(conditions)) {
        console.log(conditions);
        return {
            $and: conditions
        };
    }
    return null;
}

function getQueryOptions(userOptions) {
    var queryOptions = _.pick(userOptions, ['limit', 'skip']);
    //默认按照最后活动时间倒序排列
    queryOptions.sort = {lastActiveTime: -1};
    return queryOptions;
}
