// tracker/model/log.js
var mongoose = require('mongoose'),
    Schema = mongoose.Schema,
    ObjectId = Schema.ObjectId;

var Log = Schema({

    //the date of the log
    date: {type: Date},

    //start time of log
    start: {type: Date},

    //end time
    end: {type: Date},

    //log content
    content: String,

    //log class
    classes: Array,

    tags: Array,

    //project refer
    project: {type: ObjectId, ref: 'Project'},

    //task refer
    task: {type: ObjectId, ref: 'Task'},

    //subTask
    //subTask: {type: String, ref: 'NewTask'},

    //version
    version: {type: ObjectId, ref: "Version"},

    //signs array
    signs: Array,

    peoples: Array,

    //orignal log content
    origin: String,

    len: Number,

    //progress of project or task
    progress: {type: Object, default: null},

    //estimated time, time unit is minute
    estimatedTime: {type: Number, default: null},

    estimateStart: {type: Date, default: null},

    estimateEnd: {type: Date, default: null}
});

module.exports = mongoose.model('Log', Log);
