// tracker/exporter.js
/**
 * export data
 */
'use strict';

var Search = require('./search/search');
var Q = require('q');
var Moment = require('moment');
var _ = require('lodash');
var Msg = require('./message');


exports.export = function (options) {
    var deferred = Q.defer();
    Search.query(options).then(function (logs) {
        var days = [],
            oneDayOriginLogs = [],
            logsEndIndex = logs.length - 1,
            currentDate = null,
            prevDate = null;
        //category logs into it's belong day
        logs.forEach(function (log, index) {
            var date = new Moment(log.date).format('YYYY-MM-DD');
            currentDate = date;
            var isSameDay = index > 0 && currentDate === prevDate;
            var str = getExportLogString(log);
            if (index === 0) {
                oneDayOriginLogs.push(str);
            } else if (isSameDay) {
                oneDayOriginLogs.push(str);
            }
            //if day over, then push the data into days array
            var isDayOver = (prevDate !== null && prevDate !== currentDate ||
                index === logsEndIndex);
            if(isDayOver){
                days.push({
                    date: prevDate || currentDate,
                    content: oneDayOriginLogs.join('\n')
                });
                //initialize the next day
                oneDayOriginLogs = [getExportLogString(log)];
            }
            prevDate = currentDate;
        });
        deferred.resolve(days);
    }).catch(function (err) {
        Msg.error('exportor error!', err);
    });
    return deferred.promise;
};


function getExportLogString(log) {
    var task = log.task, subTask, parent;
    if (task) {
        parent = task.parent;
        if (parent) {
            subTask = task;
            task = parent;
        }
    }
    var project = log.project
    var progress = log.progress,
        taskPg,
        subTaskPg,
        projectPg;
    if (progress) {
        taskPg = progress.task;
        subTaskPg = progress.subTask;
        projectPg = progress.project;
    }
    var str = [
        getTimeStr(log.start, log.end),
        getLogClassStr(log.classes),
        getTagStr(log.tags),
        getSignStr(log.signs),
        getProjectStr(project, {progress: projectPg}),
        getVersionStr(log.version),
        getTaskStr(task, {progress: taskPg}),
        getSubTaskStr(subTask, {progress: subTaskPg}),
        log.content || ''
    ].filter(function (val) {
        return !!val;
    }).join(' ');
    return str;

    function getTimeStr (start, end) {
        var timeFormat = 'HH:mm';
        start =  new Moment(start);
        end = new Moment(end);
        return start.diff(end) !== 0 ? [
            start.format(timeFormat),
            end.format(timeFormat)
        ].join('~') : start.format(timeFormat);
    }

    function getLogClassStr(classes) {
        if (_.isEmpty(classes)) {
            return '';
        } else {
            return '{' + classes.map(function (cls) {
                return cls.code;
            }).join(',') + '}';
        }
    }

    function getTagStr(tags) {
        var str;
        if (_.isEmpty(tags)) {
            str =  '';
        } else {
            str = '[' + tags.join(',') + ']';
        }
        return str;
    }

    function getSignStr(signs) {
        var str;
        if (_.isEmpty(signs)) {
            str =  '';
        } else {
            str = '`' + signs.join(',') + '`';
        }
        return str;
    }


    function getVersionStr(version, props) {
        return getObj(version, '$', '$', props);
    }

    function getProjectStr(project, props) {
        return getObj(project, '<', '>', props);
    }

    function getTaskStr (task, props) {
        return getObj(task, '(', ')', props);
    }

    function getSubTaskStr(subTask, props) {
        return getObj(subTask, '#', '#', props);
    }

    function getObj(item, prefix, suffix, props) {
        if (!item) {
            return '';
        }
        var attrs = getAttrsStr(item.attributes, props);
        var tpl = _.template(prefix + '<%=name%><%=attrs ? ":" + attrs : ""%>' + suffix);
        var str = tpl({
            name: item.name,
            id: item._id.toString(),
            attrs: attrs
        });
        return str;
    }

    function getAttrsStr(attrs, props) {
        attrs = _.extend({}, attrs, props);
        if (_.isEmpty(attrs)) {
            attrs = '';
        } else {
            attrs = _.map(attrs, function (value, key) {
                if (value === undefined) {
                    return '';
                }
                return key + '="' + value + '"';
            }).join(' ');
        }
        return attrs;
    }
}
