// tracker/message.js
'use strict';

require('colors');
var msgCfg = require('./conf/config').get('message');
var EMPTY_FUN = function () {};
var _ = require('lodash');

function createMsg(level, msg) {
    return '[' + level + ']  ' + msg;
}

var api = {
    warn: function (msg) {
        console.log(createMsg('Warn', msg).yellow);
    },

    success: function(msg) {
        console.log(createMsg('Success', msg).green);
    },

    info: function(msg) {
        console.log(createMsg('Info', msg).white);
    },

    debug: function (msg) {
        console.log(createMsg('Debug', msg).grey);
    },

    log: function (msg) {
        console.log(msg);
    },

    error: function (msg, err) {
        console.log(createMsg('Error', msg).red);
        if (err) {
            console.error(err);
            console.error(err.stack);
        }
    }
};

var msgLevel = msgCfg.level,
    exportApi;
if (_.isArray(msgLevel)) {
    exportApi = {};
    _.each(api, function (fun, apiName) {
        exportApi[apiName] = _.contains(msgLevel, apiName) ? fun : EMPTY_FUN;
    });
} else {
    exportApi = api;
}

//fs.readFile(DATA_FILE_PRIFIX + argv[1]);
module.exports = exportApi;
