// tracker/util.js
'use strict';

var fs = require('fs'),
    path = require('path'),
    moment = require('moment'),
    extend = require('node.extend'),
    when = require('when');
var config = require('./conf/config.json');
var Msg = require('./message');
var Q = require('q');
//const
var DATA_FILE_PRIFIX = config.logDir;

function isValidDate(date) {
    if (Object.prototype.toString.call(date) === '[object Date]') {
        return !isNaN(date.getTime());
    } else {
        return false;
    }
}

function readLogFiles(date) {
    date = new moment(date).format('YYYY-MM-DD');
    var dateArr = parseDate(date);
    var fileName = dateArr.join('/') + '.md';
    var deferred = when.defer(),
        filePath = path.resolve(__dirname, [DATA_FILE_PRIFIX, fileName].join('/'));
    fs.readFile(filePath, 'utf8', function (err, data) {
        if (err) {
            if (err.code === 'ENOENT') {
                deferred.resolve({
                    date: date,
                    fileContent: ''
                });
            } else {
                return deferred.reject({
                    err: err,
                    message: date + '\'s log file is not exist',
                    date: date
                });
            }
        }
        if (!data) {
            Msg.warn(date + '\'s log file ' + filePath +' is exist, but the content is empty');
        }
        deferred.resolve({
            fileContent: data,
            date: date
        });
    });

    return deferred.promise;
}

function parseDate(date) {
    return date.split('-').map(function (val) {
        return parseInt(val, 10);
    });
}

function readFile(filePath, encode) {
    var deferred = when.defer();
    filePath = path.resolve(__dirname, filePath);
    fs.readFile(filePath, encode || 'utf8', function (err, data) {
        if (err) {
            return deferred.reject(err, filePath);
        }
        deferred.resolve(data);
    });
    return deferred.promise;
}

function readFileSync(filePath, encode) {
    filePath = path.resolve(__dirname, filePath);
    return fs.readFileSync(filePath, encode || 'utf8');
}

function appendFile(filePath, data) {
    var deferred = when.defer();
    filePath = path.resolve(__dirname, filePath);
    fs.appendFile(filePath, data, function (err) {
        if (err) {
            return deferred.reject(err, filePath);
        }
        deferred.resolve(data);
    });
    return deferred.promise;
}

function readLogFilesSync(date) {

    var dateArr = parseDate(date);
    var fileName = dateArr.join('/') + '.md',
        filePath = path.resolve(__dirname, [DATA_FILE_PRIFIX, fileName].join('/'));
    var fileData = fs.readFileSync(filePath, 'utf8');
    return {
        data: fileData,
        date: dateArr.join('-')
    };
}

function getDayNumInMonth (year, month) {
    return new Date(year, month, 0).getDate();
}


function isDayValid(y, m, d) {
    //the day number of one month
    var dayNum = getDayNumInMonth(y, m);
    return d <= dayNum && d > 0;
}

var dateFormat = 'YYYY-MM-DD HH:mm',
    timeFormat = 'HH:mm';
function formatDate (date) {
    return format(date, dateFormat);
}

function format (date, formatStr) {
    var m;
    if (typeof date === 'string') {
        m = new moment(date, dateFormat);
    } else {
        m = new moment(date);
    }
    return m.format(formatStr);
}

function formatTime (time) {
    return format(time, timeFormat);
}

function mean(days, prop) {
    var total = days.length;
    var sum = days.reduce(function (sum, value) {
        return sum += value[prop];
    }, 0);
    return sum/total;
}


function frequence(data, filter, creator) {
    return data.reduce(function (result, d){
        var target = result.filter(filter.bind(null, d));
        if (target && target.length > 0) {
            target[0].frequence++;
        } else {
            if (creator) {
                d = creator(d);
            }
            d = extend({frequence: 1}, d);
            result.push(d);
        }
        return result;
    }, []);
}

function inversObj(obj) {
    var result = {};
    Object.keys(obj).forEach(function (key) {
        result[obj[key]] = key;
    });
    return result;
}


function resolvePath(p) {
    return path.resolve(__dirname, p);
}




var mkdirp = require('mkdirp');
var logDir = path.normalize(config.logDir);

if (logDir.lastIndexOf('/') === logDir.length - 1) {
    logDir = logDir.slice(0, logDir.length - 1);
}
var resolvedPath = path.resolve(logDir);
var isAbsolutePath = resolvedPath === logDir;
var ext = 'md', logsPath;
if (isAbsolutePath) {
    logsPath = logDir + '/';
} else {
    logsPath = path.resolve(__dirname, '../' + logDir) + '/';
}


function writeLogFile(date, content) {
    var deferred = Q.defer();
    var dateArr = date.split('-').map(function(val) {
        return parseInt(val);
    });
    var path = logsPath + dateArr.slice(0, 2).join('/');
    //mkdir is the directory is not exist;
    mkdirp(path, function(err) {
        if (err) {
            deferred.reject(err);
            return;
        }
        var file = path + '/' + dateArr[2] + '.' + ext;
        fs.writeFile(file, content, function(err) {
            if (err) {
                Msg.error('写文件' + file + '错误', err);
                deferred.reject(err);
            }
            deferred.resolve();
        });
    });
    return deferred.promise;
}

module.exports = {
    isValidDate: isValidDate,
    readFile: readFile,
    readFileSync: readFileSync,
    appendFile: appendFile,
    readLogFiles: readLogFiles,
    readLogFileByDate: readLogFiles,
    resolvePath: resolvePath,
    readLogFilesSync: readLogFilesSync,
    writeLogFile: writeLogFile,
    getDayNumInMonth: getDayNumInMonth,
    isDayValid: isDayValid,
    formatTime: formatTime,
    formatDate: formatDate,
    timeFormat: timeFormat,
    dateFormat: dateFormat,
    frequence: frequence,
    inversObj: inversObj,
    mean: mean
};
