// ltt-addLog/Log.js
/**
 * @jsx React.DOM
 */

var React = require('react');
var $ = require('jquery');
var Ltt = global.Ltt;
var _ = require('lodash');
var TagsInput = require('./react-tagsinput');
var Moment = require('moment');

var logClasses = [
    { value: 'SPR', text : '体育'},
    { value: 'WK', text : '工作'},
    { value: 'NT', text : '一般事务'},
    { value: 'STU', text : '学习'},
    { value: 'BRK', text : '休息'},
    { value: 'TK', text : '思考'}
];
var CC_ENTER = 13,
    CC_ESC = 27;

var Tag = React.createClass({
    render: function () {
        return (
            <span className="_c-log-tag">{this.props.children}</span>
        );
    }
});

var FiltableSelect = React.createClass({
    render: function () {
        var config = _.extend({}, this.props.config);
        var value = config.value || 'value';
        var text = config.text || 'text';
        var defaultOption = {};
        defaultOption[value] = '';
        defaultOption[text] = 'select';
        options = [defaultOption].concat(this.props.options);
        var className = _.compact(["_c-filtableSelect", this.props.name]).join(' ');
        return (
            <select className={className} onChange={this.props.onChange}>
                {
                    options.map(function (opt) {
                        return (<option value={opt[value]}>{opt[text]}</option>);
                    })
                }
            </select>
        );
    }
});
var Log = React.createClass({

    getInitialState: function () {
        return {
            projects: [],
            versions: [],
            tasks: [],
            name: null,
            projectId: null,
            versionId: null,
            taskId: null
        };
    },

    render: function () {
        var projects = this.state.projects,
            versions = this.state.versions,
            tasks = this.state.tasks;
        return (
            <div className="_c-log">
                <input className="_c-log-name" placeholder="untitled task name" onKeyDown={this.onKeyDown} ref="logContent"/>
                <FiltableSelect name="logClasses" options={logClasses} ref="logClass"/>
                <FiltableSelect name="projects" options={projects} onChange={this.onProjectChange} config={{value: '_id', text: 'name'}}/>
                <FiltableSelect name="versions" options={versions} onChange={this.onVersionChange} config={{value: '_id', text: 'name'}}/>
                <FiltableSelect name="tasks" options={tasks} onChange={this.onTaskChange} config={{value: '_id', text: 'name'}}/>
                <TagsInput ref="tags"/>
                <TagsInput ref="signs"/>
            </div>
        );
    },

    componentDidMount: function () {
        var $el = $(this.getDOMNode());
        //focus on the input
        $el.find('._c-log-name').focus();
        var that = this;
        Ltt.sdk.projects().then(function (projects) {
            that.setState({
                projects: projects
            });
        });
    },

    onProjectChange: function (e) {
        var projectId = e.target.value;
        var that = this;
        if (projectId) {
            this.setState({
                projectId: projectId
            });
            Ltt.sdk.versions({projectId: projectId}).then(function (versions) {
                that.setState({
                    versions: versions
                });
            });
        }
    },

    onVersionChange: function (e) {
        var versionId = e.target.value;
        var that = this;
        if (versionId) {
            this.setState({
                versionId: versionId
            });
            Ltt.sdk.tasks({projectId: this.state.projectId, versionId: versionId})
                .then(function (tasks) {
                    that.setState({
                        tasks: tasks
                    });
                });
        }
    },

    onKeyDown: function (e) {
        var keyCode = e.keyCode;
        if (keyCode === CC_ENTER) {
            this.createLog();
        }
        if (keyCode === CC_ESC) {
            var $input = $(this.refs.logContent.getDOMNode());
            $input.blur();
        }
    },

    createLog: function () {
        var tags = this.refs.tags.getTags(),
            signs = this.refs.signs.getTags();
        console.log(tags);
        var log = _.extend({
            content: this.refs.logContent.getDOMNode().value,
            tags: tags,
            classes: [this.refs.logClass.value],
            signs: signs,
            start: new Moment().toDate()
        }, _.pick(this.state, ['projectId', 'versionId', 'taskId']));
        console.log(log);
        /*Ltt.sdk.createLog(log).then(function (log) {
            alert('创建成功');
        });*/
    }
});

module.exports = Log;
